File: ruby/lesson 3/linear_probing_hash_table.rb
class LinearProbingHashTable 
    INITIAL_CAPACITY = 4

    attr_reader :keys, :values

    def initialize(capacity = INITIAL_CAPACITY)
        @count = 0
        @capacity = capacity
        @keys = Array.new(capacity)
        @values = Array.new(capacity)
    end

    def put(key, value)
        resize(2 * @capacity) if @count >= @capacity / 2

        i = get_hash(key)
        until @keys[i].nil?
            break if key == @keys[i]
            i = (i + 1) % @capacity
        end

        @keys[i] = key
        @values[i] = value
        @count += 1
    end

    def get(key)
        i = get_hash(key)
        until @keys[i].nil?
            return @values[i] if key == @keys[i]
            i = (i + 1) % @capacity
        end
        nil
    end

    def remove(key)
        i = get_hash(key)
        until @keys[i].nil?
            if key != @keys[i]
                i = (i + 1) % @capacity
                next
            end
            @keys[i] = nil
            @values[i] = nil
            # rehash all keys in same cluster
            i = (i + 1) % @capacity
            until @keys[i].nil?
                key_to_rehash = @keys[i]
                val_to_rehash = @values[i]
                @keys[i] = nil
                @values[i] = nil
                @count -= 1
                put(key_to_rehash, val_to_rehash)
                i = (i + 1) % @capacity
            end
            @count -= 1
            break
        end
        resize(@capacity / 2) if @count.positive? && @count < @capacity / 8
    end

    private

    def get_hash(key)
        key.hash % @capacity
    end

    def resize(size)
        temp = LinearProbingHashTable.new(size)
        for i in 0..@capacity
            next if @keys[i].nil?
            temp.put(@keys[i], @values[i])
        end
        @keys = temp.keys
        @values = temp.values
        @capacity = size
    end
end

File: csharp/lesson 3/LinearProbingHashTable.cs
// https://leetcode.com/problems/design-hashmap/

public class LinearProbingHashTable<TKey, TValue> : HashTable<TKey, TValue> 
    where TKey: struct 
    where TValue: struct
{
    private const int INITIAL_CAPACITY = 4;
    private int count;
    private int capacity;
    private TKey?[] keys;
    private TValue?[] values;

    public LinearProbingHashTable(int capacity = INITIAL_CAPACITY)
    {
        this.count = 0;
        this.capacity = capacity;
        this.keys = new TKey?[capacity];
        this.values = new TValue?[capacity];
    }

    public void Put(TKey key, TValue val)
    {
        if (count >= capacity / 2) Resize(2 * capacity);
        int i;
        for (i = GetHash(key); keys[i] != null; i = (i + 1) % capacity)
        {
            if (keys[i]!.Equals(key)) break;
        }
        keys[i] = key;
        values[i] = val;
        count++;        
    }

    public TValue? Get(TKey key) {
        for (var i = GetHash(key); keys[i] != null; i = (i + 1) % capacity)
        {
            if (keys[i]!.Equals(key)) return values[i];
        }
        return null;
    }

    public void Remove(TKey key) {
        for (var i = GetHash(key); keys[i] != null; i = (i + 1) % capacity) {
            if (!keys[i]!.Equals(key)) continue;
            keys[i] = null;
            values[i] = null;
            // rehash all keys in same cluster
            i = (i + 1) % capacity;
            while (keys[i] != null) 
            {
                var keyToRehash = keys[i];
                var valToRehash = values[i];
                keys[i] = null;
                values[i] = null;
                count--;
                Put((TKey)keyToRehash, (TValue)valToRehash);
                i = (i + 1) % capacity;
            }                    
            count--;
            break;
        }
        if (count > 0 && count <= capacity / 8) Resize(capacity / 2);
    }

    private int GetHash(TKey key) 
    {
        return (key.GetHashCode() & 0x7fffffff) % capacity;
    }

    private void Resize(int size) 
    {
        // rehash all keys
        var temp = new LinearProbingHashTable<TKey, TValue>(size);
        for (var i = 0; i < capacity; i++) 
        {
            if (keys[i] == null) continue;
            temp.Put((TKey)keys[i], (TValue)values[i]);
        }
        keys = temp.keys;
        values = temp.values;
        capacity = size;
    }
}
